No more freezing if pcap process goes wrong

pboothe opened the pull request:
If packet capture fails to start for whatever reason, then the whole client gets hung in a read.  When enough clients are hung, then all incoming clients get queued, and the queue never gets any smaller.

This adds a `select()` in front of the `read()` to ensure that the pipe becomes readable within a second. Thus, if the packet trace is unable to start, the test can still proceed without it.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/ndt/51)
<!-- Reviewable:end -->

mtlynch commented:
I felt like the commit description on 5bce8cd was the most helpful in giving context for this PR. Can we integrate that into the PR description? I'm not sure if you're planning to do that as part of squash & merge, but I feel like it's helpful to aim for the final PR description as early as possible so the reviewer's getting the same context that someone would get when they go back to read this (squashed) commit in the change history.

---

Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/test_c2s_srv.c, line 533 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6fvT9cQp4FM0qwS8g:-KH6fvT9cQp4FM0qwS8h:1447350343) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L533)):*
This function is obscenely long already. We're adding 3 major chunks, so I think each of them should be reduced to (at most) single line function calls with the logic in a helper function.

(optional) If possible, we should make the net effect on `test_c2s`'s size negative (e.g. the next 30ish lines seem refactorable into a function even though we're not otherwise touching the middle 10 lines).

---

*[src/test_c2s_srv.c, line 699 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6isSE7zQorocqFmLe:-KH6isSE7zQorocqFmLf:2052795240) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L699)):*
Can we update this comment?

---

*[src/test_c2s_srv.c, line 723 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6iHgjimzXu9qLagLp:-KH6iHgjimzXu9qLagLq:1174192185) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L723)):*
Can we refactor to avoid duplicating 704-708?

---

*[src/test_c2s_srv.c, line 727 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6oGo7oEFzqEZXp5vz:-KH6oGo8L0QCfRAI1g6u:1185787727) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L727)):*
(not your fault, but while we're touching this) It seems like we reach this block when `select` times out. Is that intentional? Both `msgretvalue` and `errno` should be `0` in that case. If this is intentional, can we document that, because it's a bit unintuitive.

---

*[src/test_c2s_srv.c, line 738 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6pjnVuBcWo73B3MgI:-KH6pjnVuBcWo73B3MgJ:1181294136) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L738)):*
Was this referring to what is now 725?

---

*[src/test_c2s_srv.c, line 766 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6r1Rh4GlYxEBF8Jqz:-KH6r1Rh4GlYxEBF8Jr-:508835245) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L766)):*
Can we document why we don't care if this blocks?

---

*[src/test_c2s_srv.c, line 767 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6r3u56q6pVE1BvmJZ:-KH6r3u56q6pVE1BvmJ_:-772321278) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L767)):*
Suggest `packet_trace_running = 0` for good measure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/51#-:-KH6rttLmcNj_AYCdUBh:-890946271)*
<!-- Sent from Reviewable.io -->

pboothe commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/test_c2s_srv.c, line 533 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6fvT9cQp4FM0qwS8g:-KH719k9b_yJ6sMIz3mp:-1055957585) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L533)):*
I agree that this function is a dumpster fire. However, bugfixes should not be held up by cleanups.

Aside from the fact that refactoring a 530 line function is not what I signed up for as fixer of this bug,  cleanups entail more risk than I think we can tolerate at this time. We need this bug to be fixed ASAP, and fixed in a way that definitely doesn't introduce any new bugs. Every cleanup I have ever done of code in NDT has revealed hidden assumptions and dependencies between far-flung pieces of the program, and has broken a lot of things that worked by accident. I don't want that to happen here and prevent the timely rollout of a fix.

---

*[src/test_c2s_srv.c, line 699 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6isSE7zQorocqFmLe:-KH70ncPXh5iBZBAAMsg:1836917993) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L699)):*
I have no idea what this comment ever meant. I can delete it, but can not update it.

---

*[src/test_c2s_srv.c, line 723 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6iHgjimzXu9qLagLp:-KH70xMqD3y5HZ7BlIuW:-1220765007) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L723)):*
You're proposing spending 12 lines to avoid spending 4, so I'd prefer to keep as is. If we start engaging in cleanups we will be here for weeks.

---

*[src/test_c2s_srv.c, line 727 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6oGo7oEFzqEZXp5vz:-KH6zlaW755ibqwvmVDC:-20940249) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L727)):*
It is intentional, and is a preservation of existing behavior. Documented.

---

*[src/test_c2s_srv.c, line 738 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6pjnVuBcWo73B3MgI:-KH70PU-WaWTq3xoNu8X:-547965660) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L738)):*
It was and is referring to the `read()` on the next line.

---

*[src/test_c2s_srv.c, line 766 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6r1Rh4GlYxEBF8Jqz:-KH7-E3CTSvFTgw7Kkrx:1251061640) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L766)):*
I am preserving existing behavior, and I have no idea whether later pieces of code will reliably work okay if I change this IO to non-blocking.

The only reason it's okay to change the other one to non-blocking is because the alternative is never calling `stop_packet_trace()` at all, and so we mitigate the risk of resource leaks by calling it in non-blocking mode as part of an obscure error pathway.

I would rate changing the blocking/non-blocking status of this particular write as a risky change. Every time I have changed a piece of NDT code in a way that affects threading and/or IPC it has caused bugs, sometimes a lot and in code that is very distant, and that's what the blocking/non-blocking change affects.

---

*[src/test_c2s_srv.c, line 767 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6r3u56q6pVE1BvmJZ:-KH7-474SeECSA1Wkka0:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L767)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/51)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/test_c2s_srv.c, line 533 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6fvT9cQp4FM0qwS8g:-KHKtL4C4YBuh9i6M9Oz:20382196) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L533)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> I agree that this function is a dumpster fire. However, bugfixes should not be held up by cleanups.
> 
> Aside from the fact that refactoring a 530 line function is not what I signed up for as fixer of this bug,  cleanups entail more risk than I think we can tolerate at this time. We need this bug to be fixed ASAP, and fixed in a way that definitely doesn't introduce any new bugs. Every cleanup I have ever done of code in NDT has revealed hidden assumptions and dependencies between far-flung pieces of the program, and has broken a lot of things that worked by accident. I don't want that to happen here and prevent the timely rollout of a fix.</details>
I think we might be talking about different things. When I say that we should add single function calls instead of adding code blocks, I'm talking about something like changing this block to something like:

```
int wait_for_packet_trace_start(packet_trace_pipe) {
   ...
}

...

int test_c2s(...) {
   ...
   packet_trace_running = wait_for_packet_trace_start(mon_pipe[0]);
   ...
}
```

This isn't a cleanup because we're not touching existing code. We're just adding code in a way that minimizes stress on an overtaxed function.

The suggestion I marked as optional *does* involve lightly touching existing code, but that's why I marked it as optional.

---

*[src/test_c2s_srv.c, line 699 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6isSE7zQorocqFmLe:-KHKvQt2rtc54ba9rAWS:-1718874844) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L699)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> I have no idea what this comment ever meant. I can delete it, but can not update it.</details>
I was talking about s/speed-chk/packet tracing/ but if it makes more sense to delete entirely, that's fine.

---

*[src/test_c2s_srv.c, line 723 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6iHgjimzXu9qLagLp:-KH6nB0jrI0RA798znTM:-556600093) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L723)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> You're proposing spending 12 lines to avoid spending 4, so I'd prefer to keep as is. If we start engaging in cleanups we will be here for weeks.</details>
704-708 were fine prior to this change (modulo magic numbers) but if we copy/paste them, we're introducing a problem. I see this as mess prevention rather than cleanup of existing code.

Can we move 704-708 to within the `for` loop? That doesn't cause a net increase in LOC.

If not, I think minimizing the bloat of `test_c2s` is more important than minimizing overall LOC of the file, so I think adding a function is worth it even if it causes the overall LOC to increase.

That said, I don't understand the "spending 12 lines to avoid spending 4" calculation. We could refactor the 4 lines into a function that's 6-7 lines, then we reduce each 4-line block to a 1-line function call, so it seems like spending 7 lines to save 6. To me, this seems desirable considering that the 6 lines are coming out of a gigantic function.

---

*[src/test_c2s_srv.c, line 738 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6pjnVuBcWo73B3MgI:-KHKzMlIgUJA-BEv1CU4:1747650641) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L738)):*
> ```C
>       } else {
>         /* There is something to read, so get it from the pktpair child.  If an
>          * interrupt occurs, just skip the read and go on
> ```

I'm confused then. It says:

>If an interrupt occurs, just skip the read and go on

But the interrupt handling seems to be above and not here. Is `read` internally handling interrupts?

---

*[src/test_c2s_srv.c, line 766 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6r1Rh4GlYxEBF8Jqz:-KHKzcaa87-awlBGHtZl:-1825920510) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L766)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> I am preserving existing behavior, and I have no idea whether later pieces of code will reliably work okay if I change this IO to non-blocking.
> 
> The only reason it's okay to change the other one to non-blocking is because the alternative is never calling `stop_packet_trace()` at all, and so we mitigate the risk of resource leaks by calling it in non-blocking mode as part of an obscure error pathway.
> 
> I would rate changing the blocking/non-blocking status of this particular write as a risky change. Every time I have changed a piece of NDT code in a way that affects threading and/or IPC it has caused bugs, sometimes a lot and in code that is very distant, and that's what the blocking/non-blocking change affects.</details>
I'm not sure if my note came across clearly. I'm not proposing that we change behavior here; I'm asking that we document why we've introduced inconsistency between the two `stop_packet_trace` calls.

---

*[src/test_c2s_srv.c, line 181 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHKrAYj0nUfMZKRxBTf:-KHKrAYj0nUfMZKRxBTg:649073664) ([raw file](https://github.com/m-lab/ndt/blob/bb6d79cd0849bc993c441b865b2186401715d9c6/src/test_c2s_srv.c#L181)):*
> ```C
>     activeStreams = connections_to_fd_set(c2s_conns, streamsNum, &rfd, &max_fd);
>   }
> }
> ```

I think this got overlooked because it was a top-level comment, so repeating it in code:

>I felt like the commit description on 5bce8cd was the most helpful in giving context for this PR. Can we integrate that into the PR description? I'm not sure if you're planning to do that as part of squash & merge, but I feel like it's helpful to aim for the final PR description as early as possible so the reviewer's getting the same context that someone would get when they go back to read this (squashed) commit in the change history.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/51)*
<!-- Sent from Reviewable.io -->

pboothe commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/test_c2s_srv.c, line 533 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6fvT9cQp4FM0qwS8g:-KHLP8v9IoBUMGKy1git:-1787878844) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L533)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> I think we might be talking about different things. When I say that we should add single function calls instead of adding code blocks, I'm talking about something like changing this block to something like:
> 
> ```
> int wait_for_packet_trace_start(packet_trace_pipe) {
>    ...
> }
> 
> ...
> 
> int test_c2s(...) {
>    ...
>    packet_trace_running = wait_for_packet_trace_start(mon_pipe[0]);
>    ...
> }
> ```
> 
> This isn't a cleanup because we're not touching existing code. We're just adding code in a way that minimizes stress on an overtaxed function.
> 
> The suggestion I marked as optional *does* involve lightly touching existing code, but that's why I marked it as optional.</details>
Made the helper function for this chunk, left the optional suggestion alone. Although later changes do actually get close to that goal.

---

*[src/test_c2s_srv.c, line 699 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6isSE7zQorocqFmLe:-KHL2pZ-tRzPjo67cAV-:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L699)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> I was talking about s/speed-chk/packet tracing/ but if it makes more sense to delete entirely, that's fine.</details>
Done.

---

*[src/test_c2s_srv.c, line 723 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6iHgjimzXu9qLagLp:-KHL3Jeuj2p_cENHwf2w:1634540702) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L723)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> 704-708 were fine prior to this change (modulo magic numbers) but if we copy/paste them, we're introducing a problem. I see this as mess prevention rather than cleanup of existing code.
> 
> Can we move 704-708 to within the `for` loop? That doesn't cause a net increase in LOC.
> 
> If not, I think minimizing the bloat of `test_c2s` is more important than minimizing overall LOC of the file, so I think adding a function is worth it even if it causes the overall LOC to increase.
> 
> That said, I don't understand the "spending 12 lines to avoid spending 4" calculation. We could refactor the 4 lines into a function that's 6-7 lines, then we reduce each 4-line block to a 1-line function call, so it seems like spending 7 lines to save 6. To me, this seems desirable considering that the 6 lines are coming out of a gigantic function.</details>
Moved inside, which actually allowed the deletion of several lines at the end.

---

*[src/test_c2s_srv.c, line 738 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6pjnVuBcWo73B3MgI:-KHL1GLyFYAZWdkRzs9D:-591796295) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L738)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> I'm confused then. It says:
> 
> >If an interrupt occurs, just skip the read and go on
> 
> But the interrupt handling seems to be above and not here. Is `read` internally handling interrupts?</details>
No, instead if `read()` is interrupted it will return -1 and set `errno` to `EINTR`.  I read this comment as an explanation for why it is okay to not have a `continue` or fancy flow control in this `if (...) {` block.

The error handling above this line deals with errors that happened during `select()` not any that may happen during `read()`.

---

*[src/test_c2s_srv.c, line 766 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6r1Rh4GlYxEBF8Jqz:-KHL1fCEUDuaOkVOw4DV:-783605252) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L766)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> I'm not sure if my note came across clearly. I'm not proposing that we change behavior here; I'm asking that we document why we've introduced inconsistency between the two `stop_packet_trace` calls.</details>
Cool. But the reason is "preserve existing behavior", which is a strange justification to put in a comment. A comment saying `// This is like this because it used to be like this` feels a bit like [a sign warning about the sharp edges on the sign](http://www.shutterstock.com/pic-55017349/stock-photo-stupid-sign-warning-sign-has-sharp-edges.html). I will do it if you really think it aids readability, but if I encountered such a comment I would not thank the author for writing it.

---

*[src/test_c2s_srv.c, line 181 \[r2\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHKrAYj0nUfMZKRxBTf:-KHL3BmLYT95Alxr2irU:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/bb6d79cd0849bc993c441b865b2186401715d9c6/src/test_c2s_srv.c#L181)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> I think this got overlooked because it was a top-level comment, so repeating it in code:
> 
> >I felt like the commit description on 5bce8cd was the most helpful in giving context for this PR. Can we integrate that into the PR description? I'm not sure if you're planning to do that as part of squash & merge, but I feel like it's helpful to aim for the final PR description as early as possible so the reviewer's getting the same context that someone would get when they go back to read this (squashed) commit in the change history.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/51)*
<!-- Sent from Reviewable.io -->

mtlynch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> I felt like the commit description on 5bce8cd was the most helpful in giving context for this PR. Can we integrate that into the PR description? I'm not sure if you're planning to do that as part of squash & merge, but I feel like it's helpful to aim for the final PR description as early as possible so the reviewer's getting the same context that someone would get when they go back to read this (squashed) commit in the change history.</details>


---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/test_c2s_srv.c, line 766 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6r1Rh4GlYxEBF8Jqz:-KHLfr-zP2TSfTbC8NO1:1473190120) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L766)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary>
> Cool. But the reason is "preserve existing behavior", which is a strange justification to put in a comment. A comment saying `// This is like this because it used to be like this` feels a bit like [a sign warning about the sharp edges on the sign](http://www.shutterstock.com/pic-55017349/stock-photo-stupid-sign-warning-sign-has-sharp-edges.html). I will do it if you really think it aids readability, but if I encountered such a comment I would not thank the author for writing it.</details>
I feel like TODOs are good notices of "we know this is weird, but we haven't fixed it yet." Suggest:

```c
// TODO: Refactor to use shutdown_packet_trace_non_blocking once we verify that
// blocking is not necessary.
```

---

*[src/test_c2s_srv.c, line 198 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLgTxrrCZL9PgqWDGg:-KHLgTxrrCZL9PgqWDGh:385252898) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L198)):*
> ```C
>  * non-blocking. This should not affect any code because any shutdown messages
>  * sent should be the last usage of the pipe anyway.
>  * @param mon_pipe the pipe to send on which to send shutdown messages.
> ```

extra "to send"

---

*[src/test_c2s_srv.c, line 200 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLbQEHW-td7IQQ-IKR:-KHLbQEHW-td7IQQ-IKS:282386706) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L200)):*
> ```C
>  * @param mon_pipe the pipe to send on which to send shutdown messages.
>  */
> void packet_trace_emergency_shutdown(int *mon_pipe) {
> ```

(optional) I have no idea what the `mon` in `mon_pipe` means. French developer? We have an opportunity to give it a clearer name here, but at the same time I see the value in using the same variable name throughout.

---

*[src/test_c2s_srv.c, line 200 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLfJrOUvoLFK_EW3dP:-KHLfJrPilyEjM-kODRd:-1160155063) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L200)):*
> ```C
>  * @param mon_pipe the pipe to send on which to send shutdown messages.
>  */
> void packet_trace_emergency_shutdown(int *mon_pipe) {
> ```

Suggest `shutdown_packet_trace_non_blocking`. Emergency seems to assume intent rather than describe functionality. A caller might want to shut down packet tracing and just not block (emergency or not).

---

*[src/test_c2s_srv.c, line 215 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLcUGG82bO4ojRHoGQ:-KHLcUGG82bO4ojRHoGR:-1550068561) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L215)):*
> ```C
> /** Waits up to one second for the passed-in fd to become readable.
>  * @param fd the file descriptor to wait for
>  * @returns true if the fd is readable, false otherwise
> ```

(optional) Do we want to say "true" and "false"? It seems like the rest of the comments are in terms of 1 and 0.

---

*[src/test_c2s_srv.c, line 223 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLeFB2dwYH2NekjPzc:-KHLeFB3AL89GYqsgPQo:1714724500) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L223)):*
> ```C
>   FD_SET(fd, &rfd);
>   memset(&sel_tv, 0, sizeof(sel_tv));
>   sel_tv.tv_sec = 1;  // Wait for up to 1 second
> ```

(optional) I feel like we've talked about this at some point, and I don't recall if there's a deliberate reason not to do this in NDT, but the previous six lines could be reduced to:

```c
fd_set rfd = {};
struct timeval sel_tv = {};
FD_SET(fd, &rfd);
sel_tv.tv_sec = 1;  // Wait for up to 1 second
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/51#-:-KHLhCCEBP81Jtr2nGRT:1417088721)*
<!-- Sent from Reviewable.io -->

pboothe commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/test_c2s_srv.c, line 766 \[r1\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KH6r1Rh4GlYxEBF8Jqz:-KHMD4JHz8aERDMT986a:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/b9b94fe18adf534f359bfcae157cbd463340f841/src/test_c2s_srv.c#L766)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> I feel like TODOs are good notices of "we know this is weird, but we haven't fixed it yet." Suggest:
> 
> ```c
> // TODO: Refactor to use shutdown_packet_trace_non_blocking once we verify that
> // blocking is not necessary.
> ```</details>
Done.

---

*[src/test_c2s_srv.c, line 198 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLgTxrrCZL9PgqWDGg:-KHM4PzsY4zEmEivUYJJ:-499146009) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L198)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> extra "to send"</details>
Done.

---

*[src/test_c2s_srv.c, line 200 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLbQEHW-td7IQQ-IKR:-KHM4RGmKwSqOKaiVMSu:-1451908695) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L200)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> (optional) I have no idea what the `mon` in `mon_pipe` means. French developer? We have an opportunity to give it a clearer name here, but at the same time I see the value in using the same variable name throughout.</details>
monitoring pipe, I have always assumed. Keeping the name because the function is single-use. 

---

*[src/test_c2s_srv.c, line 200 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLfJrOUvoLFK_EW3dP:-KHM4gN05PUrDrlmHyiy:1584223545) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L200)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> Suggest `shutdown_packet_trace_non_blocking`. Emergency seems to assume intent rather than describe functionality. A caller might want to shut down packet tracing and just not block (emergency or not).</details>
I like "emergency" because it allows the error path to be `log_println(0, ...)` instead of having to write a reasonable error path like one would do for a function that was not meant to be called as a last resort. If this function were meant to be called in the normal course of events, I would write it differently and it would have a different signature.

---

*[src/test_c2s_srv.c, line 215 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLcUGG82bO4ojRHoGQ:-KHM59lPfciHLrI9144c:1648309472) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L215)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> (optional) Do we want to say "true" and "false"? It seems like the rest of the comments are in terms of 1 and 0.</details>
I did, because I return the result of a boolean statement, so I return a truth value, which is really "non-zero" and "zero" according to the C spec and I just don't want to even get into it so I said "true" and "false" to avoid the discussion of whether boolean operators are guaranteed to always return 1 for true.

---

*[src/test_c2s_srv.c, line 223 \[r3\]](https://reviewable.io:443/reviews/m-lab/ndt/51#-KHLeFB2dwYH2NekjPzc:-KHMD4nnHToSJoIOiETj:-921298692) ([raw file](https://github.com/m-lab/ndt/blob/fcd88a0eb96ad7204ae961b039a0487c01d26533/src/test_c2s_srv.c#L223)):*
<details><summary><i>Previously, mtlynch (Michael Lynch) wrote…</i></summary>
> (optional) I feel like we've talked about this at some point, and I don't recall if there's a deliberate reason not to do this in NDT, but the previous six lines could be reduced to:
> 
> ```c
> fd_set rfd = {};
> struct timeval sel_tv = {};
> FD_SET(fd, &rfd);
> sel_tv.tv_sec = 1;  // Wait for up to 1 second
> ```</details>
Pedantry note:

```C
fd_set rfd = {0};
struct timeval sel_tv = {0};
FD_SET(fd, &rfd);
sel_tv.tv_sec = 1;  // Wait for up to 1 second
```
Also: good call. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt/51)*
<!-- Sent from Reviewable.io -->
